nntrainer/tensor/tensor.cpp

```diff
@@ -2148,1680 +734,401 @@ Tensor &Tensor::dot_batched_deriv_wrt_2(Tensor &m_deriv,
-    throw std::invalid_argument("Error: Tensor::cos supports fp32 case only.");
+void Tensor::print(std::ostream &out) const {
+  printInstance(out, this);
+  itensor->print(out);
 }
 
-void Tensor::inv_sqrt_i() {
-  if (getDataType() == ml::train::TensorDim::DataType::FP32) {
-    if (!contiguous) {
-      apply_i<float>([](float val) -> float { return 1 / std::sqrt(val); });
-    } else {
-      inv_sqrt_inplace(this->size(), getData<float>());
-    }
-  } else if (getDataType() == ml::train::TensorDim::DataType::FP16) {
-#ifdef ENABLE_FP16
-    if (!contiguous) {
-      apply_i<_FP16>([](_FP16 val) -> _FP16 {
-        return static_cast<_FP16>(1 / std::sqrt(static_cast<float>(val)));
-      });
-    } else {
-      inv_sqrt_inplace(this->size(), getData<_FP16>());
-    }
-#else
-    throw std::invalid_argument("Error: enable-fp16 is not enabled");
-#endif
-  } else
-    throw std::invalid_argument(
-      "Error: Tensor::inv_sqrt_i only supports fp32, fp16");
-}
+void Tensor::putData() const { itensor->putData(); }
 
-float Tensor::l2norm() const {
-  NNTR_THROW_IF(!contiguous, std::invalid_argument)
-    << getName() << " is not contiguous, cannot get l2norm.";
-  float ret = 0;
-  unsigned int len = size();
-  if (getDataType() == ml::train::TensorDim::DataType::FP32) {
-    const float *data = getData<float>();
-    ret = snrm2(len, data, 1);
-  } else if (getDataType() == ml::train::TensorDim::DataType::FP16) {
-#ifdef ENABLE_FP16
-    const _FP16 *data = getData<_FP16>();
-    ret = snrm2(len, data, 1);
-#else
-    throw std::invalid_argument("Error: enable-fp16 is not enabled");
-#endif
+void Tensor::setData(const std::shared_ptr<MemoryData> buf, size_t off,
+                     bool init) {
+  itensor->setMemoryData(buf, off);
+
+  if (buf && init) {
+    initialize();
   }
-  return ret;
 }
 
-float Tensor::max_abs() const {
-  NNTR_THROW_IF(!contiguous, std::invalid_argument)
-    << getName() << " is not contiguous, cannot get max_abs.";
+const std::shared_ptr<MemoryData> Tensor::getMemoryData() const {
+  return itensor->getMemoryData();
+}
 
-  unsigned int len = size();
-  float ret = 0;
-  if (getDataType() == ml::train::TensorDim::DataType::FP32) {
-    const float *data = getData<float>();
+size_t Tensor::getOffset() const { return itensor->getOffset(); }
 
-    unsigned int idx = isamax(len, data, 1);
-    ret = *(data + idx);
+void Tensor::copy(const Tensor &from) {
+  /// @todo enable copy to non-contiguous tensor
+  if (!itensor->getContiguous()) {
+    throw std::runtime_error("Cannot copy non-contiguous tensor");
+  }
 
-  } else if (getDataType() == ml::train::TensorDim::DataType::FP16) {
+  if (from.size() != 0 && size() == from.size() &&
+      getDataType() == from.getDataType()) {
+    // if tensor size and data type match, copy data
+    itensor->copy(from);
+  } else {
+    // replace with a new tensor that are the same with the given tensor
+    if (from.getDataType() == ml::train::TensorDim::DataType::FP32) {
```

Comment:
Can we remove this if branch by calling itensor.copy?

Reply:
I left this code intentionally in the tensor class for the following reasons.

1. This condition checks the data type of the Tensor to copy. Therefore, even if this code is moved into Float/HalfTensor, we still need to check the source Tensor's data type, which eventually has the same code in both classes.
2. Current implementation uses swap() to perform copy. However, this won't be a valid option when it happens in the Float/HalfTensor class. This is because `swap(t, *this)` would be `swap(Tensor &lhs, FloatTensor &rhs)` or `swap(Tensor &lhs, HalfTensor &rhs)` instead of `swap(Tensor &lhs, Tensor &rhs)`. While Tensor has FloatTensor as a member variable,  I believe implementing a swap between Tensor and FloatTensor would require structural change.

Hope this clarifies and let me know if you have further questions.

Reply:
Also, if you have any suggestions, please bring them up!

Reply:
how about using template? Cause if we need to add more tensor type like int8, bf16, then there will be more branch.

Reply:
On second thought, we can just remove this branch and create Tensor T using copy constructor.

Reply:
fixed!